### PR TITLE
bugfix for #1853: garbage collector deletes guest users

### LINKED ISSUE #1853: An eager GC deletes guest users while getting created
Prevent GC from deleting guest users while they are being created and assigned a project.

### Tentative solution
- Context manager around study creation that locks GC from deleting a given user. Then when redirection is returned, extend lock a timeout until client connects again the server and the standard TTL process starts.
- As a temporary solution, GC operation interval has been increased in master and staging. P2E shows better results.

### PR DESCRIPTION
**Pull request by pcrespov**
## What do these changes do?

Access to ``/study/{study_id}`` link produces
image

The reason is that the Garbage collector removes GUEST users when they have no resources associated. This caused that temporary GUEST users created in studies access were deleted even before their projects were associated.

This fix overcomes this problem by 
- first creating the user as ANONYMOUS instead of GUEST
- then, once the ``access_study`` is processed, a fictitious socket is created for the socket and it is turn into GUEST

This will prevent the GC to destroy the user before the ``access_study`` handle is completed and some extra time TTL to reconnect.

- [x] Adapted ``pytest tests/unit/with_dbs/fast/test_access_to_studies.py``
- [x] test to demo that an eager GC does not delete user
- cleaned

## Related issue number

Fixes #1853

## How to test
```cmd
cd services/web
make install-dev
pytest --pdb -s --log-cli-level=DEBUG tests/unit/with_dbs/fast/test_access_to_studies.py
```
and can also see GC log.

## Checklist

- [x] Did you change any service's API? Then make sure to bundle document and upgrade version (``make openapi-specs``, ``git commit ...`` and then ``make version-*``)
- [x] Unit tests for the changes exist
- [x] Runs in the swarm
- [x] Documentation reflects the changes
- [x] New module? Add your github username to [.github/CODEOWNERS](.github/CODEOWNERS)

### CODE REVIEW

File: services/web/server/src/simcore_service_webserver/application.py [hunk 1]
```diff
@@ -68,14 +68,16 @@ def create_application(config: Dict) -> web.Application:
...
     setup_projects(app)  # needs storage
-    setup_studies_access(app)
     setup_activity(app)
     setup_resource_manager(app)
     setup_tags(app)
     setup_catalog(app)
     setup_publications(app)
     setup_products(app)
 
+    # top level plugins
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/application.py:78]: what is a top level plugin vs what?

File: services/web/server/src/simcore_service_webserver/studies_access.py [hunk 2]
```diff
@@ -55,41 +55,57 @@ async def get_public_project(app: web.Application, project_uuid: str):
...
 
+    # creates a user that is marked as ANONYMOUS. see update_user_as_guest
     user = await db.create_user(
         {
             "name": username,
             "email": email,
             "password_hash": encrypt_password(password),
             "status": ACTIVE,
-            "role": GUEST,
+            "role": ANONYMOUS,
```
Comment (sanderegg) [inline services/web/server/src/simcore_service_webserver/studies_access.py:81]: Just to understand what is happenning here. so normally it is impossible for an anonymous user to be registered right? and the GC does not check them?

**Review by sanderegg — APPROVED** [review]

File: services/web/server/src/simcore_service_webserver/studies_access.py (hunk 2 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/studies_access.py:81]: Exactly.

The user is "temporary" anonymous until it is turn into GUEST just after everything was setup correctly and before responding with a redirection
I avoid to create a GUEST from the beginning to avoid that the GC might delete it between this point and when we create the entry in redis.

File: services/web/server/src/simcore_service_webserver/studies_access.py [hunk 3]
```diff
@@ -55,41 +55,57 @@ async def get_public_project(app: web.Application, project_uuid: str):
...
+    from .resource_manager.websocket_manager import WebsocketRegistry
+
+    if user.get("role") == ANONYMOUS:
+        db = get_storage(app)
+        username = user["name"]
+
+        # creates an entry in the socket's registry to avoid garbage collector (GC) deleting it
+        # See https://github.com/ITISFoundation/osparc-simcore/issues/1853
+        registry = WebsocketRegistry(user["id"], f"{username}-guest-session", app)
+        await registry.set_socket_id(f"{username}-guest-socket-id")
```
Comment (GitHK) [inline services/web/server/src/simcore_service_webserver/studies_access.py:101]: So this will give you exactly 1 period of time, equal to the value of `resource_deletion_timeout_seconds` which defaults to 900 seconds (if was not changes on the platforms). In this period of time the user will not be removed by the GC.

This will give `resource_deletion_timeout_seconds ` seconds for the user to be logged in. Please keep in mind that there is a mechanism where if the user pressed logout, its resources will be collected even faster. Search for `# trigger faster gc on disconnect`.

If I remember correctly e2e also presses logout when it finishes with a test.

**Review by GitHK — COMMENTED** [review]
Please try to understand if there are any corner cases from what I've commented below.

File: services/web/server/src/simcore_service_webserver/application.py (hunk 1 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/application.py]: not sure myself :-) removed it

File: services/web/server/src/simcore_service_webserver/studies_access.py (hunk 3 above)
Comment (pcrespov) [reply services/web/server/src/simcore_service_webserver/studies_access.py:101]: > So this will give you exactly 1 period of time, equal to the value of `resource_deletion_timeout_seconds` which defaults to 900 seconds (if was not changes on the platforms). In this period of time the user will not be removed by the GC.

Correct. In any case, if you look in detail, in reality resource_deletion_timeout_seconds will be at least 3 secs. I can

> This will give `resource_deletion_timeout_seconds ` seconds for the user to be logged in.

Yes. But notice that that value starts counting from the moment the server responds **with a redirection** ... meaning that it should immediately 


 > Please keep in mind that there is a mechanism where if the user pressed logout, its resources will be collected even faster. Search for `# trigger faster gc on disconnect`.

That is fine. A guest user that logs out should be GC-ed

**Review by GitHK — APPROVED** [review]
This should address the issue, due to the fact that the users are being used immediately after creation.
